multi-component-workload tutorial creation (#29)
karmada-io/playground: karmada-multi-component-workload-example/step6/text.md

### Provide Workload Definitions to Karmada

Before Karmada can schedule complex FlinkDeployment workloads, it needs to understand their structure.

> To achieve this, we must apply their **Custom Resource Definitions (CRDs)** to the Karmada control plane and propagate them to the member clusters.

Reviewer: ```suggestion
To achieve this, we must apply their **Custom Resource Definitions (CRDs)** to the Karmada control plane and propagate them to the member clusters.
In addition, Karmada uses the Resource Interpreter to parse CR information such as resource requests, dependencies, health status and more, which is critical for scheduling and orchestration. For multi-component scheduling scenarios, Karmada primarily relies on the Resource Interpreter mechanism to extract precise resource requirements at the component granularity. 
For more information on customizing the Karmada Resource Interpreter, please refer to https://karmada.io/docs/next/userguide/globalview/customizing-resource-interpreter
```

> > **Note:** Karmada (v1.17+) has built-in support for interpreting FlinkDeployment workloads. It automatically handles extraction of each component's replicas and resource requirements, so no manual Resource Interpreter is needed.

Reviewer: ```suggestion
> **Note:** Karmada has built-in support for interpreting FlinkDeployment workloads. It automatically handles extraction of each component's replicas and resource requirements, so no manual Resource Interpreter is needed.
```

Actually, it starts from v1.15. But we can simply remove this info

### Provide Workload Definitions to Karmada

Before Karmada can schedule complex FlinkDeployment workloads, it needs to understand their structure. 

To achieve this, we must do two things:
1. Apply their **Custom Resource Definitions (CRDs)** to the Karmada control plane and propagate them to the member clusters.
2. Apply **Resource Interpreter Customizations** to teach Karmada how to extract per-component resource requirements from these specific workload types.

> **Note:** Karmada has built-in support for interpreting common third-party multi-component workload resources such as FlinkDeployment. They define rules for Karmada to parse these resources, covering extraction of replicas and resource requirements of each component, judgment of workload health status and identification of dependent resources.

**Apply the CRDs and PropagationPolicy:**

RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config apply --validate=false -f /root/examples/flinkdeployments.flink.apache.org-v1.yaml`{{exec}}

> This applies the Flink CRD.

Reviewer: ```suggestion
This applies the FlinkDeployment CRD.
```

Please use the full name

<details>
<summary>crd-propagation-policy.yaml</summary>

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: ClusterPropagationPolicy
metadata:
  name: crd-propagation
spec:
  resourceSelectors:
    - apiVersion: apiextensions.k8s.io/v1
      kind: CustomResourceDefinition
      name: flinkdeployments.flink.apache.org
  placement:
    clusterAffinity:
      clusterNames:
        - kind-member1
        - kind-member2
```

</details>

RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config apply --validate=false -f /root/examples/crd-propagation-policy.yaml`{{exec}}

This applies the ClusterPropagationPolicy to distribute the CRDs.

**Apply the Resource Interpreters:**

Karmada needs to know exactly how to parse the custom resources to find their component definitions. We provide Lua scripts that teach Karmada how to do this.

> RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config apply -f /root/examples/flink-interpreter.yaml`{{exec}}

Reviewer: We have built-in interpreters, why do we need apply the interpreter here?

Author: You're right, thanks for catching that, I will just fix it

### Provide Workload Definitions to Karmada

Before Karmada can schedule complex FlinkDeployment workloads, it needs to understand their structure.

To achieve this, we must do two things:

1. Apply their **Custom Resource Definitions (CRDs)** to the Karmada control plane and propagate them to the member clusters.
2. Apply **Resource Interpreter Customizations** to teach Karmada how to extract per-component resource requirements from these specific workload types.

> **Note:** Karmada has built-in support for interpreting common third-party multi-component workload resources such as FlinkDeployment. They define rules for Karmada to parse these resources, covering extraction of replicas and resource requirements of each component, judgment of workload health status and identification of dependent resources.

**Apply the CRDs and PropagationPolicy:**

> RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config apply --validate=false -f /root/examples/flinkdeployments.flink.apache.org-v1.yaml`{{exec}}

Reviewer: I still have the question: why do we need `validate=false` here?

This applies the FlinkDeployment CRD.

<details>
<summary>crd-propagation-policy.yaml</summary>

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: ClusterPropagationPolicy
metadata:
  name: crd-propagation
spec:
  resourceSelectors:
    - apiVersion: apiextensions.k8s.io/v1
      kind: CustomResourceDefinition
      name: flinkdeployments.flink.apache.org
  placement:
    clusterAffinity:
      clusterNames:
        - kind-member1
        - kind-member2
```

</details>

RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config apply -f /root/examples/crd-propagation-policy.yaml`{{exec}}

This applies the ClusterPropagationPolicy to distribute the CRDs.

**Apply the Resource Interpreters:**

> While Karmada's newer versions have built-in support for parsing FlinkDeployment workloads, the version installed in this environment requires us to explicitly provide a Lua script that teaches Karmada how to extract the component definitions.

Reviewer: > the version installed in this environment requires us to explicitly provide a Lua script that teaches Karmada how to extract the component definitions.

The karmadactl version is `v1.17.2`, it should have the resource interpter for `FlinkDeployment` . So why should we explicitly provide a Lua script again?

Author: I ran into the null components issue during testing , I reverted the changes but I was actually caused by something else , I have fixed it now

### Provide Workload Definitions to Karmada

> Before Karmada can schedule complex Flink and Volcano workloads, it needs to understand their structure. 

Reviewer: ```suggestion
Before Karmada can schedule complex FlinkDeployment workloads, it needs to understand their structure. 
```

Using FlinkDeployment for the demonstration is sufficient.

To achieve this, we must do two things:
1. Apply their **Custom Resource Definitions (CRDs)** to the Karmada control plane and propagate them to the member clusters.
2. Apply **Resource Interpreter Customizations** to teach Karmada how to extract per-component resource requirements from these specific workload types.

> > **Note:** We have pre-downloaded the necessary CRDs and placed them in `/root/examples/` for you.

Reviewer: ```suggestion
> **Note:** Karmada has built-in support for interpreting common third-party multi-component workload resources such as FlinkDeployment and VolcanoJob. They define rules for Karmada to parse these resources, covering extraction of replicas and resource requirements of each component, judgment of workload health status and identification of dependent resources.
```

      kind: CustomResourceDefinition
      name: flinkdeployments.flink.apache.org
    - apiVersion: apiextensions.k8s.io/v1
      kind: CustomResourceDefinition
      name: jobs.batch.volcano.sh
  placement:
    clusterAffinity:
      clusterNames:
        - kind-member1
        - kind-member2
```

</details>

RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config apply --validate=false -f /root/examples/crd-propagation-policy.yaml`{{exec}}

This applies the ClusterPropagationPolicy to distribute the CRDs.

**Apply the Resource Interpreter Customizations:**

Karmada uses a built-in "Resource Interpreter" to dynamically inspect unfamiliar custom resources. By applying these Lua-based configurations, we teach the interpreter exactly where to look in a `FlinkDeployment` and `VolcanoJob` to find their individual components, replicas, and CPU/Memory requests.

RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config apply -f /root/examples/flink-interpreter.yaml`{{exec}}

This applies the Flink Resource Interpreter Customization.

RUN `kubectl --kubeconfig /etc/karmada/karmada-apiserver.config apply -f /root/examples/volcano-interpreter.yaml`{{exec}}

> This applies the Volcano Resource Interpreter Customization.

Reviewer: ```suggestion
```
